crates/runtime/src/query_context.rs

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{sync::Arc, time::SystemTime};
+
+use arrow::datatypes::Schema;
+use datafusion::{
+    execution::SendableRecordBatchStream, physical_plan::stream::RecordBatchStreamAdapter,
+};
+use tokio::sync::RwLock;
+
+use crate::{datafusion::DataFusion, query_history};
+use async_stream::stream;
+use futures::StreamExt;
+
+pub struct QueryContext {
+    pub query_id: uuid::Uuid,
+    pub schema: Option<Arc<Schema>>,
+    pub sql: Option<String>,
+    pub nsql: Option<String>,
+    pub start_time: SystemTime,
+    pub end_time: Option<SystemTime>,
+    pub execution_time: Option<u64>,
+    pub rows_produced: Option<u64>,
```

Comment:
Instead of using options for all these, I would purely use integers/enums. E.g.

`results_cache_hit`

- `-1` not set
- `0` miss
- `1` hit

`rows_produced`

- `-1` not set
- etc.

Reply:
Like it 👍 propose to add this change as a separate iteration/PR to reduce num of changes introduced. I'll improve this as next step including query `error` indication (don't currently have this).